### ISSUE TITLE
[sklearn#34622] explained_variance_score returns a perfect 1.0 for a single sample (silently breaks LeaveOneOut CV)

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#308]
**Mirrored from [scikit-learn#34622](https://github.com/scikit-learn/scikit-learn/issues/34622)**

Original title: explained_variance_score returns a perfect 1.0 for a single sample (silently breaks LeaveOneOut CV)
Original labels: Needs Triage
Created: 2026-08-01T04:15:36Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the bug

`explained_variance_score` returns `1.0` (a perfect score) for a single sample no
matter how wrong the prediction is, and emits no warning. `r2_score`, its
documented sibling, guards the same degenerate case with an
`UndefinedMetricWarning` and returns `nan`.

With one sample there is no variance to explain, so the numerator
(`Var(y_true - y_pred)`) and the denominator (`Var(y_true)`) are both `0`, and
`force_finite` maps `0/0` to `1.0` — documented as "perfect predictions" — even
when the prediction is plainly wrong.

The practical consequence is that **`LeaveOneOut` cross-validation scored with
`explained_variance` silently reports a perfect 1.0 for any model, including one
fit on pure noise.**

### Steps/Code to Reproduce

```python
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import explained_variance_score, r2_score
from sklearn.model_selection import LeaveOneOut, cross_val_score

# 1) single sample, clearly wrong prediction
print(explained_variance_score([1.0], [2.0]))  # 1.0, no warning
print(r2_score([1.0], [2.0]))                  # nan + UndefinedMetricWarning

# 2) the practical impact: LeaveOneOut on pure noise
rng = np.random.RandomState(0)
X, y = rng.normal(size=(12, 3)), rng.normal(size=12)

ev = cross_val_score(LinearRegression(), X, y, cv=LeaveOneOut(),
                     scoring="explained_variance")
r2 = cross_val_score(LinearRegression(), X, y, cv=LeaveOneOut(), scoring="r2")
print(ev.mean())  # 1.0   <- perfect score on noise, silently
print(r2.mean())  # nan   <- correctly undefined
```

### Expected Results

Something that does not claim a perfect fit on a single sample. Matching
`r2_score` seems most consistent: emit `UndefinedMetricWarning` and return `nan`
when `n_samples < 2`, since the metric is not defined without variance to
explain.

### Actual Results

```
1.0        # explained_variance_score([1.0], [2.0]) -- no warning
nan        # r2_score([1.0], [2.0]) -- UndefinedMetricWarning

1.0        # LeaveOneOut + explained_variance on pure noise
nan        # LeaveOneOut + r2 on pure noise (12 UndefinedMetricWarnings)
```

### Versions

```
System:
    python: 3.14
Python dependencies:
    sklearn: 1.9.0
```

### Notes

I appreciate that ignoring a systematic offset is intentional for this metric
(the docstring notes it), so the `n_samples >= 2` behaviour is not in question
here — only the single-sample case, where there is no variance at all and the
score is silently reported as perfect.

I'

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/115 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
